### DISCUSSION TITLE
mongodb

### DISCUSSION THREAD

**Discussion by prohorsev** [phalcon/cphalcon#15466] [Q&A] [answered]
Tell me please how to use mongodb? I installed phalcon, devtools, incubator 3.4, mongodb 4.4. Where can I insert settings (dbname, login, password), config.php? How should I use incubator adapter, services.php? Where can I insert php class Collection, which extends MongoCollection, in models? Or any other ways?

**Comment by Jeckerson** [comment]
Hello, `v3.4` is deprecated version, use `v4.1` instead.

Also use updated mongo-db composer package - https://github.com/phalcon/incubator-mongodb
`composer require phalcon/incubator-mongodb`

Code example:
```php
$di->setShared('mongo', function () {
        $mongo = new Client("mongodb://mongo-db-url:27017");

        return $mongo->selectDatabase("hello_world");
    });

    // Registering the mongoDB CollectionManager service
    $di->setShared('collectionsManager', function () {
        return new MongoDBCollectionManager();
    });
```

Collection
```php
<?php

use Phalcon\Incubator\MongoDB\Mvc\Collection;

class UsersCollection extends Collection
{
    public function onConstruct(): void
    {
        $this->setSource('users');
        $this->useImplicitObjectIds(false);
    }
}
```

Controller
```php
<?php

use Phalcon\Mvc\Controller;
use Phalcon\Mvc\View;
use Phalcon\Mvc\Model\Resultset;

class MongoController extends Controller
{
    public function indexAction()
    {
        return UsersCollection::findFirst([
            'conditions' => [
                '_id' => mt_rand(1, 10000),
            ],
            'projection' => [
                //'id' => 0,
            ],
        ]);
    }
```

**Reply by prohorsev** [reply]
Hello. Insert it in services.php:
```php
$di->setShared('mongo', function () {
        $mongo = new Client("mongodb://mongo-db-url:27017");

        return $mongo->selectDatabase("test");
    });

    // Registering the mongoDB CollectionManager service
    $di->setShared('collectionsManager', function () {
        return new MongoDBCollectionManager();
    });
```
Client and MongoDBCollectionManager are unknown classes.
Insert it in library\test\TestCollection.php:
```php
<?php

use Phalcon\Incubator\MongoDB\Mvc\Collection;

class TestCollection extends Collection
{
    public function onConstruct(): void
    {
        $this->setSource('TestCollection');
        $this->useImplicitObjectIds(false);
    }
}
```
Insert it in MongoController.php:
```php
<?php

use Phalcon\Mvc\Controller;
use Phalcon\Mvc\View;
use Phalcon\Mvc\Model\Resultset;

class MongoController extends Controller
{
    public function indexAction()
    {
        return TestCollection::findFirst(['conditions' => [
            '_id' => mt_rand(1, 10000),
        ],
            'projection' => [
                //'id' => 0,
            ],]);
    }
}
```
Go to page /mongo. Error 500. What's wrong?
Mongodb:
![image](https://user-images.githubusercontent.com/55758955/117366835-01196380-aeca-11eb-81de-95fc3d80bab7.png)
app:
![image](https://user-images.githubusercontent.com/55758955/117367072-535a8480-aeca-11eb-9653-975acbd15ee2.png)

**Comment by Jeckerson** [comment, marked as answer]
> Client and MongoDBCollectionManager are unknown classes.

https://github.com/TechEmpower/FrameworkBenchmarks/blob/9a5b198b0b054fa068d3f9e6e0e9f9fb40a4b71a/frameworks/PHP/phalcon/public/index.php#L8

You can see example here.

**Reply by prohorsev** [reply]
Thank you. Excellent repository. How can I debug mongodb connection? It still doesn't work. Should I enable access control via installing mongodb server https://docs.mongodb.com/manual/tutorial/enable-authentication/ ?

**Reply by prohorsev** [reply]
Problem was:
- composer require mongodb/mongodb
- isert in index.php: 
```php
include __DIR__ . '/../vendor/autoload.php';
```
